### PR TITLE
WithInfantryBody now supports playing "make" animation

### LINKED ISSUE #12632: Deploy animation doesn't play with "WithInfantryBody"
In RA mod, I've inherited from default ^Soldier class.

	WithInfantryBody:
		DefaultAttackSequence: shoot
		UpgradeTypes: undeployed
		UpgradeMinEnabledLevel: 1
	WithSpriteBody:
		# Need this for deploy animation.
	WithIdleOverlay@DEPLOYED:
		Sequence: deploy-shoot
		UpgradeTypes: deployed
		UpgradeMinEnabledLevel: 1
	DeployToUpgrade:
		DeployedUpgrades: deployed
		UndeployedUpgrades: undeployed
		CanDeployOnRamps: true
		DeployAnimation: deploy

However, DeployAnimation didn't work for me until I added WithSpriteBody trait, as ^Soldier class only have WithInfantryBody. It is not documented in traits document that DeployAnimation is only read when WithSpriteBody is present.

I think this is affecting D2K mod thumper. I don't see its deploy animation.

### PR DESCRIPTION
**Pull request by forcecore**
Much needed animated infantries with RA2 in mind. (and currently affects D2K thumper)

Since WithSpriteBody and WithInfantryBody have quite a different animation management, it seems suitable to introduce an IPlayCustomAnimation interface for those that provide custom animation logics. Sounds sensible, it has potential to support animated voxel deployer.

To test this, run D2K mod then deploy a thumper. Unfortunately, thumper deploy animation (make sequence in thumper entry) is only 5 frames long and is too short to see. Extend its length for testing purpose if needed. Currently, thumper can move while the lengthy deploy/undeploy animation is playing. That is probably due to GrantConditionOnDeploy not taking away mobility while the animation is playing. (or is it the YAML not using Undeployed for the mobile condition?)

Demo: https://www.youtube.com/watch?v=o4jyIDtD72s

Fixes #12632

### CODE REVIEW

**Comment by forcecore** [comment]
Disallowed attacking animation to interrupt custom animation. Otherwise the unit will be stuck in "deploying" state and won't be able to deploy ever again.

**Comment by forcecore** [comment]
Corrected YAML for thumper to disallow it from moving while deploying/undeploying

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 1]
```diff
@@ -443,4 +443,15 @@ public override string ValueChangedMessage(string playerName, string newValue)
 			return playerName + " " + BoolValues[newValue] + " " + Name + ".";
 		}
 	}
+
+	public interface IPlayCustomAnimation
+	{
+		void PlayCustomAnimation(Actor self, string name, Action after = null);
+		void PlayCustomAnimationRepeating(Actor self, string name);
+		void PlayCustomAnimationBackwards(Actor self, string name, Action after = null);
+		void PlayFetcnIndex(Actor self, string name, Func<int> func);
```
Comment (reaperrr) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]: Typo: should be 'Fetc**h**'

File: OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs [hunk 2]
```diff
@@ -99,14 +99,14 @@ public void Reverse(Actor self, Activity activity, bool queued = true)
...
+				var pca = pcas.FirstOrDefault(Exts.IsTraitEnabled);
 
 				// HACK: The actor remains alive and active for one tick before the followup activity
 				// (sell/transform/etc) runs. This causes visual glitches that we attempt to minimize
 				// by forcing the animation to frame 0 and regranting the make condition.
 				// These workarounds will break the actor if the followup activity doesn't dispose it!
-				if (wsb != null)
-					wsb.DefaultAnimation.PlayFetchIndex(info.Sequence, () => 0);
+				if (pca != null)
+					pca.PlayFetcnIndex(self, info.Sequence, () => 0);
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs]: Fetc**h**

File: OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs [hunk 3]
```diff
@@ -130,5 +134,10 @@ void INotifyDamageStateChanged.DamageStateChanged(Actor self, AttackInfo e)
 		{
 			DamageStateChanged(self);
 		}
+
+		public void PlayFetcnIndex(Actor self, string name, Func<int> func)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs]: Fetc**h**

**Comment by atlimit8** [comment]
Please squash the typo fix into the first commit.

**Comment by forcecore** [comment]
Typo fixed and squashed

**Comment by forcecore** [comment]
A thought about "Disallow attacking animation to interrupt Deploy animation" commit. I think it is the modders problem to sort this out: Armament shouldn't use "!deployed" condition instead of the correct "undeployed" condition. It will look odd in the mod but that should be better than making modders go "Help! I can't make this infantry deployed!". Fortunately, and unfortunately for the modders who seek examples from official mods, Thumpers don't have armament.

**Comment by forcecore** [comment]
Can you put this timing diagram in GrantConditionOnDeploy trait documentation?
https://github.com/OpenRA/OpenRA/wiki/Traits#grantconditionondeploy

deploy svg

**Comment by forcecore** [comment]
Retracting this.
Doesn't work on bleed/playtest anymore even though the merge was a success. D2K infantry can be fixed without this anyway.

**Pull request closed without merging**